app/src/main/java/org/mozilla/fenix/home/recentvisits/RecentVisitsFeature.kt

```diff
@@ -0,0 +1,258 @@
+import org.mozilla.fenix.home.recentvisits.RecentlyVisitedItem.RecentHistoryGroup
+import org.mozilla.fenix.home.recentvisits.RecentlyVisitedItem.RecentHistoryHighlight
+import kotlin.math.max
+
+@VisibleForTesting internal const val MAX_RESULTS_TOTAL = 9
+@VisibleForTesting internal const val MIN_VIEW_TIME_OF_HIGHLIGHT = 10.0
+@VisibleForTesting internal const val MIN_FREQUENCY_OF_HIGHLIGHT = 4.0
+
+/**
+ * View-bound feature that retrieves a list of [HistoryHighlight]s and [HistoryMetadata] items
+ * which will be mapped to [RecentlyVisitedItem]s and then dispatched to [HomeFragmentStore]
+ * to be displayed on the homescreen.
+ *
+ * @param homeStore The [HomeFragmentStore] that holds the state of the [HomeFragment].
+ * @param historyMetadataStorage The storage that manages [HistoryMetadata].
+ * @param historyHighlightsStorage The storage that manages [PlacesHistoryStorage].
+ * @param scope The [CoroutineScope] used for IO operations related to querying history
+ * and then for dispatching updates.
+ * @param ioDispatcher The [CoroutineDispatcher] for performing read/write operations.
+ */
+class RecentVisitsFeature(
+    private val homeStore: HomeFragmentStore,
+    private val historyMetadataStorage: HistoryMetadataStorage,
+    private val historyHighlightsStorage: Lazy<PlacesHistoryStorage>,
+    private val scope: CoroutineScope,
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
+) : LifecycleAwareFeature {
+
+    private var job: Job? = null
+
+    override fun start() {
+        job = scope.launch(ioDispatcher) {
+            val historyHighlights = async {
+                historyHighlightsStorage.value.getHistoryHighlights(
+                    HistoryHighlightWeights(MIN_VIEW_TIME_OF_HIGHLIGHT, MIN_FREQUENCY_OF_HIGHLIGHT),
+                    MAX_RESULTS_TOTAL
+                )
+            }
+
+            val allHistoryMetadata = async {
+                historyMetadataStorage.getHistoryMetadataSince(Long.MIN_VALUE)
+            }
+
+            // val dedupedHistoryHighlights = dedupeVisits(historyHighlights.await(), allHistoryMetadata.await())
+
+            val featureHighlights = getFeatureHighlights(historyHighlights.await(), allHistoryMetadata.await())
+            val featureGroups = getFeatureGroups(allHistoryMetadata.await())
+
+            updateState(featureHighlights, featureGroups)
+        }
+    }
+
+    /**
+     * Do an in-memory query for the metadata of history highlights to get each item's last accessed date.
+     *
+     * Safe to call:
+     *   - if [metadata] doesn't contain the details for all [highlights].
+     *   The highlights for which the last access tiem is not found will have it set to a default of `0`.
+     *   - if the same [HistoryHighlight.url] appears multiple times in [metadata].
+     *   The latest time of any of these urls was last accessed will be associated to the highlight.
+     *
```

Comment:
nit: 

If a [highlight] cannot be mapped to a corresponding [metadata] record, its `lastAccessTime` will be set to 0.
If a [highlight] maps to multiple metadata records, its `lastAccessTime` will be set to the most recently updated record.